Add more documentation for `$` (#533)
sindresorhus/execa: docs/scripts.md

await Promise.all([
  $`sleep 1`,
  $`sleep 2`,
  $`sleep 3`,
]);

const dirName = 'foo bar';
await $`mkdir /tmp/${dirName}`;
```

## Summary

This file describes the differences between Bash, Execa, and [zx](https://github.com/google/zx) (which inspired this feature).

### Flexibility

Unlike shell languages like Bash, libraries like Execa and zx enable you to write scripts with a more featureful programming language (JavaScript). This allows complex logic (such as [parallel execution](#parallel-commands)) to be expressed easily. This also lets you use [any Node module](#builtin-utilities).

### Shell

The main difference between Execa and zx is that Execa does not require any shell. Shell-specific keywords and features are [written in JavaScript](#variable-substitution) instead.

This is more cross-platform. For example, your code works the same on Windows machines without Bash installed.

Also, there is no shell syntax to remember: everything is just plain JavaScript.

If you really need a shell though, the [`shell` option](../readme.md#shell) can be used.

### Simplicity

Execa's scripting API mostly consists of only two methods: [``$`command` ``](../readme.md#command) and

> [`$(options)`](../readme.md#options).

Reviewer: Don't hard-wrap

Author: Fixed :+1:

[No special binary](#main-binary) is recommended, no [global variable](#global-variables) is injected: scripts are regular Node.js files.

Execa is a thin wrapper around the core Node.js [`child_process` module](https://nodejs.org/api/child_process.html). Unlike zx, it lets you use [any of its native features](#background-processes): [`pid`](#pid), [IPC](https://nodejs.org/api/child_process.html#subprocesssendmessage-sendhandle-options-callback), [`unref()`](https://nodejs.org/api/child_process.html#subprocessunref), [`detached`](https://nodejs.org/api/child_process.html#child_processspawncommand-args-options), [`uid`](https://nodejs.org/api/child_process.html#child_processspawncommand-args-options), [`gid`](https://nodejs.org/api/child_process.html#child_processspawncommand-args-options), [`signal`](https://nodejs.org/api/child_process.html#child_processspawncommand-args-options), etc.

### Modularity

zx includes many builtin utilities: `fetch()`, `question()`, `sleep()`, `stdin()`, `retry()`, `spinner()`, `chalk`, `fs-extra`, `os`, `path`, `globby`, `yaml`, `minimist`, `which`, Markdown scripts, remote scripts.

> Execa does not include [any utility](#builtin-utilities): it focuses on being small and modular instead. Any Node module can be used in your scripts.

Reviewer: ```suggestion
Execa does not include [any utility](#builtin-utilities): it focuses on being small and modular instead. Any Node.js package can be used in your scripts.
```

### Performance

Spawning a shell for every command comes at a performance cost, which Execa avoids.

Also, [local binaries](#local-binaries) can be directly executed without using `npx`.

### Debugging

Child processes can be hard to debug, which is why Execa includes a [`verbose` option](#verbose-mode).

Also, Execa's error messages and [properties](#errors) are very detailed to make it clear to determine why a process failed.

> Finally, unlike Bash and zx which are stateful (options, current directory, etc.), Execa is [purely functional](#current-directory), which also helps with debugging.

Reviewer: ```suggestion
Finally, unlike Bash and zx, which are stateful (options, current directory, etc.), Execa is [purely functional](#current-directory), which also helps with debugging.
```

//   #!/usr/bin/env zx
```

```js
// Execa scripts are just regular Node.js files
node file.js
```

### Global variables

```js
// zx
await $`echo example`;
```

```js
// Execa
import {$} from 'execa';
await $`echo example`;
```

### Command execution

```sh
# Bash
echo example
```

```js
// zx
await $`echo example`;
```

```js
// Execa
await $`echo example`;
```

### Subcommands

```sh
# Bash
echo "$(echo example)"
```

```js
// zx
const example = await $`echo example`;
await $`echo ${example}`;
```

```js
// Execa
const example = await $`echo example`;
await $`echo ${example}`;
```

### Parallel commands

```sh
# Bash
echo one &
echo two &
```

```js
// zx
await Promise.all([$`echo one`, $`echo two`]);
```

```js
// Execa
await Promise.all([$`echo one`, $`echo two`]);
```

### Serial commands

```sh
# Bash
echo one && echo two
```

```js
// zx
await $`echo one && echo two`;
```

```js
// Execa
await $`echo one`;
await $`echo two`;
```

### Local binaries

```sh
# Bash
npx tsc --version
```

```js
// zx
await $`npx tsc --version`;
```

```js
// Execa
await $`tsc --version`;
```

### Builtin utilities

```js
// zx
const content = await stdin();
```

```js
// Execa

> import getStdin from 'get-stdin';
> const content = await getStdin();

Reviewer: ```suggestion
import getStdin from 'get-stdin';

const content = await getStdin();
```

Author: Fixed :+1:

```

```js
// Execa
await $({env: {EXAMPLE: '1'}})`example_command`;
```

### Escaping

```sh
# Bash
echo 'one two'
```

```js
// zx
await $`echo ${'one two'}`;
```

```js
// Execa
await $`echo ${'one two'}`;
```

### Escaping multiple arguments

```sh
# Bash
echo 'one two' '$'
```

```js
// zx
await $`echo ${['one two', '$']}`;
```

```js
// Execa
await $`echo ${['one two', '$']}`;
```

### Current filename

```sh
# Bash
echo "$(basename "$0")"
```

```js
// zx
await $`echo ${__filename}`;
```

```js
// Execa
import {fileURLToPath} from "node:url";
import {basename} from "node:path";
const __filename = basename(fileURLToPath(import.meta.url));

await $`echo ${__filename}`;
```

### Verbose mode

```sh
# Bash
set -v
echo example
```

```js
// zx >=8
await $`echo example`.verbose();

// or:
$.verbose = true;
```

```js
// Execa
const $$ = $({verbose: true});
await $$`echo example`;
```

Or:

```
NODE_DEBUG=execa node file.js
```

### Current directory

```sh
# Bash
cd project
```

```js
// zx
cd('project');

// or:
$.cwd = 'project';
```

```js
// Execa
const $$ = $({cwd: 'project'});
```

### Multiple current directories

```sh
# Bash
pushd project
pwd
popd
pwd
```

```js
// zx

> within(async () => {

Author: This does not seem to be the case according to their documentation: https://github.com/google/zx/#within
Although according to their type, it would: https://github.com/google/zx/blob/4999435182fe1f34f6b2ef2471184d9878456b5e/src/core.ts#L436 :thinking:

> within(async () => {

Reviewer 2: Wonder if that's a typo. I saw that and their second example including an await/assignment.  Not a show stopper imo
